Bluetooth HCI SPI slave application

mbolivar opened the pull request:
This PR adds a sample application which can be used as the slave side of an HCI SPI solution.

It's been tested on 96b_carbon_nrf51 with some additional patches for 96b_carbon which are still being prepared for upstream submission. I don't have an nrf51_pca10028 to test with, but a project configuration file is also provided for this board, which was tested with an earlier version of this application.

I'm happy to drop this PR until the combined patches for 96Boards Carbon are complete, but since it's a useful component in its own right, I thought to submit separately.

rsalveti committed:
samples: Bluetooth: add HCI SPI raw handler sample

Add a sample application that allows a Zephyr-based Bluetooth
controller to interface with an HCI driver via SPI. This sample
implements the same BT SPI protocol already as Zephyr's HCI SPI
driver.

Currently, the sample only supports the legacy SPI API.

Provide a single configuration file, avoiding board-specific
files. Some board-specific configuration information must be provided
via other means:

- CONFIG_BT_CONTROLLER_TO_HOST_SPI_DEV_NAME
- CONFIG_BT_CONTROLLER_TO_HOST_SPI_IRQ_DEV_NAME
- CONFIG_BT_CONTROLLER_TO_HOST_SPI_IRQ_PIN

Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>
Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>
Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
Bluetooth: hci: spi: abstract out legacy SPI API

Some targets (e.g. STM32-based boards) only have SPI drivers using the
new API. To support both these and existing, legacy SPI drivers in the
SPI HCI driver, abstract out the SPI API into shim routines.

There are no behavioral differences due to this patch. The next patch
will add support for the new API.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
Bluetooth: hci: spi: add support for new SPI API

Add shim routines for the new SPI driver API.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
doc: 96Boards Carbon: add Bluetooth setup docs

Add cross-referenced information on the 96b_carbon and
96b_carbon_nrf51 pages which disambiguates between the two "boards".
Also describe how to flash 96b_carbon_nrf51 with
samples/bluetooth/hci_spi and 96b_carbon with samples/bluetooth/ipsp
to support a Bluetooth HCI stack on 96Boards Carbon (the physical
board).

While we're here, make the documentation page for 96b_carbon match the
format in doc/templates/board.tmpl.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

mbolivar commented:
@carlescufi @cvinayak @pfalcon FYI

jhedberg commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,288 @@
+		evt_hdr->evt = BT_HCI_EVT_VENDOR;
+		evt_hdr->len = 2;
+		u16_t *param = net_buf_add(buf, sizeof(u16_t));
+		*param = sys_cpu_to_le16(EVT_BLUE_INITIALIZED);
```
This looks like a potential unaligned access violation since you can't be 100% sure that the pointer returned by net_buf_add sits on a 2-byte boundary. You should instead use net_buf_add_le16(buf, EVT_BLUE_INITIALIZED).

cvinayak commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,288 @@
+#include <bluetooth/hci.h>
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_raw.h>
+#include <bluetooth/common/log.h>
```
in the hci_uart we use #include "common/log.h", i tried that here for some reason cant include and compiler throws error.

The point is common/log.h is not public, not in include/bluetooth/ folder

cvinayak commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,288 @@
+NET_BUF_POOL_DEFINE(cmd_tx_pool, CONFIG_BLUETOOTH_HCI_CMD_COUNT, CMD_BUF_SIZE,
+		    BT_BUF_USER_DATA_MIN, NULL);
+
+#define BT_L2CAP_MTU 65 /* 64-byte public key + opcode */
```
There is a conditional compilation introduced in hci_uart to pick the MTU from controller's configured tx buffer size, please use that code from hci_uart

mbolivar commented:
Thanks for the reviews; I've addressed them in v2.

Since this was first written before I joined Linaro, I also looked at the diff between the source and hci_{uart,usb}. There appear to be some structural similarities between hci_uart and hci_spi (switch/break layout, `#defines`, and included headers), so I've also added back in the Nordic and Intel copyright. I didn't see significant enough code overlap to merit adding in Signed-off-by: lines. Please let me know if that doesn't seem kosher.

cvinayak commented on samples/bluetooth/hci_spi/prj_nrf51_pca10028.conf:
```diff
@@ -0,0 +1,28 @@
+CONFIG_BLUETOOTH_DEBUG_LOG=y
+
+# SPI port and pins to communicate with the SPI master
+CONFIG_SPI=y
```
Remove, already defined in line 2

cvinayak commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,296 @@
+
+#include "common/log.h"
+
+#define HCI_CMD			0x01
```
nitpicking, general comment throughout in this file, use spaces to align macro values (instead of tabs as of now e.g. between HCI_CMD \t\t\t 0x01).

cvinayak commented on samples/bluetooth/hci_spi/prj_nrf51_pca10028.conf:
```diff
@@ -0,0 +1,28 @@
+# Debug
+CONFIG_SYS_LOG=y
+CONFIG_SYS_LOG_DEFAULT_LEVEL=2
+CONFIG_BLUETOOTH_DEBUG_LOG=y
```
with BLUETOOTH_DEBUG_LOG and SYS_LOG both enabled, I get compile warning: "SYS_LOG_LEVEL" redefined.

We dont need debug enabled in default config, please remove for these debug defines for both boards.

mbolivar commented:
Thanks for the additional review.

> We dont need debug enabled in default config, please remove for these debug defines for both boards.

I've removed them from prj_nrf51_pca10028.conf, but left them commented out in prj_96b_carbon_nrf51.conf so they're easy to re-enable later.

> Remove, already defined in line 2

Done.

> nitpicking, general comment throughout in this file, use spaces to align macro values (instead of tabs as of now e.g. between HCI_CMD \t\t\t 0x01).

Done.

dbkinder commented on samples/bluetooth/hci_spi/README.rst:
```diff
@@ -0,0 +1,19 @@
+Requirements
+************
+
+* A board with SPI and BLE support
```
Do you have an example you've tested you could list here as an example?

dbkinder commented on samples/bluetooth/hci_spi/README.rst:
```diff
@@ -0,0 +1,19 @@
+Overview
+********
+
+Expose Zephyr Bluetooth Controller support over SPI to another device/CPU using
```
This sample exposes the Zephyr Bluetooth...

dbkinder commented on samples/bluetooth/hci_spi/README.rst:
```diff
@@ -0,0 +1,19 @@
+Building and Running
+********************
+
+See :ref:`bluetooth setup section <bluetooth_setup>` for details.
```
This references explains how to setup and run Bluetooth samples under QEMU and not on actual hardware.  Need some hardware specific instructions here.

pfalcon replied:
It seems that @mbolivar just copy-pasted existing README.rst from samples/bluetooth/hci_uart/ and samples/bluetooth/hci_usb/ . So, README in this patch is as good or as bad as READMEs for those samples. And indeed, they are pretty bad:

* https://www.zephyrproject.org/doc/samples/bluetooth/hci_uart/README.html
* https://www.zephyrproject.org/doc/samples/bluetooth/hci_usb/README.html

E.g.:

> Requirements
> * BlueZ running on the host, or
> * A board with BLE support

What does that "or" do there?

mbolivar replied:
> What does that "or" do there?
Yeah, looks like it should be "and". Will fix.

pfalcon replied:
@mbolivar : Sorry for the confusion, my comment was mostly for @dbkinder , showing that your short README for hci_spi is on par with ones hci_uart and hci_usb, and *they* have issues like that (which all need to be fixed, but hopefully in separate pull requests, and not blocking this patch ;-) ).

dbkinder reviewed (changes requested):
Needs hardware specific setup and running instructions.

mbolivar commented:
> Needs hardware specific setup and running instructions.

I intend to post these in the follow-up patch series I referenced in the pull request summary (I actually have these parts written), but I can't post those yet as I'm still working on some patches for 96b_carbon (not 96b_carbon_nrf51) required to enable this.

Are you OK with deferring that to future work, or should I drop this series until both are ready?

pfalcon commented:
Built this PR, it builds ;-). Hope to try to run it next few days.

@jhedberg , @cvinayak : Your comments seems to have been addressed by @mbolivar , can you please re-review?

mbolivar commented:
> Built this PR, it builds ;-). Hope to try to run it next few days.

I hope so too! Note that I've also submitted https://github.com/zephyrproject-rtos/zephyr/pull/832. We've been using an out of tree SPI driver for 96b_carbon and I ran into some problems with the mainlined version which PR 832 fixes.

When that's merged, I'll be able to submit the "other half" of this along with hardware-specific docs for usage on 96Boards Carbon (the board, combining the 96b_carbon and 96b_carbon_nrf51 Zephyr "BOARD"s).

pfalcon commented:
@mbolivar :

> some patches for 96b_carbon (not 96b_carbon_nrf51)

To clarify, this app should be flashed to nrf51 on the Carbon, which, per previous discussion on 96boards-dev list requires: a) soldering SWD pins for nrf51; b) dedicated SWD adapter - is that true?

mbolivar commented:
> To clarify, this app should be flashed to nrf51 on the Carbon, which, per previous discussion on 96boards-dev list requires: a) soldering SWD pins for nrf51; b) dedicated SWD adapter - is that true?

On 96Boards Carbon, yes, it can be flashed onto the nrf51. It can be flashed onto other dev boards with an nrf51 (or any other SoC with a Zephyr SPI slave driver and a BT stack) as well, of course.

I do have a Carbon where I soldered on the nrf51 debug header to flash this, but I find it's not necessary. With an SWD debugger that lets me connect flying 0.1" jumper wire leads (I use the BlackMagic probe with 20 pin JTAG adapter kit and cables, https://1bitsquared.com/products/black-magic-probe), I can just insert the leads into the unpopulated header, press them into place with one hand, and type one-handed into the debugger with the other hand to flash the chip. This works well if you've got some .gdbinit functions defined to attach your debugger and run "load". Feel free to ping me on IRC if you want to discuss further.

pfalcon commented:
@mbolivar : Ok, I did my homework and was able to access Carbon's nRF51 using its STM32 running BlackMagic probe. I then was able to flash this app using:

~~~
(gdb) load zephyr.elf
Loading section text, size 0xd656 lma 0x0
Loading section .ARM.exidx, size 0x8 lma 0xd658
Loading section devconfig, size 0x84 lma 0xd660
Loading section rodata, size 0x364 lma 0xd6e4
Loading section datas, size 0x1b0 lma 0xda48
Loading section initlevel, size 0x84 lma 0xdbf8
Loading section _k_sem_area, size 0x3c lma 0xdc7c
Loading section _k_mutex_area, size 0x14 lma 0xdcb8
Loading section _k_queue_area, size 0x28 lma 0xdccc
Loading section _net_buf_pool_area, size 0x6c lma 0xdcf4
Start address 0xc268, load size 56670
Transfer rate: 16 KB/sec, 858 bytes/write.
~~~

Of course, there're no visible effects of that (how to easily run something on nRF51 with visible effect?), so I'm waiting for further instructions ;-).

pfalcon commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,296 @@
+		evt_hdr->len = 2;
+		net_buf_add_le16(buf, EVT_BLUE_INITIALIZED);
+	}
+	spi_send(buf);
```
Should this spi_send() be outside `if` above?

mbolivar replied:
I think actually due to the K_FOREVER the entire if should be deleted and the code executed unconditionally, no?

pfalcon commented:
> how to easily run something on nRF51 with visible effect?

Ok, I wasn't able to confirm that it's possible to flash/run anything on nRF51. I followed an obvious idea: make blinky run on nRF51, then run on STM32 "gpio proxy" app which "forwards" gpio signal from one pin to another (LED). I used BLE_RDY (P0.29 -> PB0). Nothing works ;-). Won't spend more time on reverse-engineering Carbon's funky design, will wait for detailed instructions ;-).

mbolivar commented:
> will wait for detailed instructions ;-).

@pfalcon, here is a preview of what's to come:

https://github.com/mbolivar/zephyr/tree/96b_carbon-net

It builds on the hci_spi app for nRF51 (this PR #803), the bug fix PR #832 I mentioned above, some additional needed Carbon patches, and the documentation @dbkinder requested tying this all together (docs patch is https://github.com/mbolivar/zephyr/commit/78e980c9e586c16da50894d81980f3d461e3dbdf)

As I mentioned, though, fixing the SPI on STM32F4 in PR #832 is a blocker before I can submit the contents of this branch upstream, and there are some issues with it that will take some remote debugging to resolve.

(Since 1. the hci_spi slave app, 2. the SPI driver bug fixes, and 3. the Carbon patches to tie things together are logically separate, I am trying to get 1. and 2. merged incrementally first to avoid confusing the review by tying too many subsystems together in one PR.)

pfalcon commented:
@mbolivar : Thanks, that branch worked for me (down to ping6'ing the Carbon).

pfalcon commented:
> down to ping6'ing the Carbon

Correction: I stupidly pinged local end of IPv6 conn. Now redoing it properly, I hit a typical "nothing works" situation, e.g. https://jira.zephyrproject.org/browse/ZEP-2422

pfalcon commented:
When I fixed something, I got correct TCP echo from ipsp sample, and then somewhere on TCP disconnect:

~~~
***** BUS FAULT *****
  Executing thread ID (thread): 0x20002080
  Faulting instruction address:  0x80009ee
  Precise data bus error
  Address: 0xe5f7ffc1
Fatal fault in essential thread! Spinning...
~~~

(Many of the recent posts aren't related directly to this patch, but as we were deprived of issue tracker here... Jira doesn't seem good for such "stream of debugging" posts.)

andrewboie commented:
>  Jira doesn't seem good for such "stream of debugging" posts

What does this even mean?
Is having JIRA open in a different browser tab and clicking the giant "Create" button at the top of the page really that burdensome?
If you're going to complain about the current bug tracking system, please give very specific, actionable feedback on why this is the case.
Otherwise, if an issue is reproducible in the master branch, file a bug...

pfalcon commented:
Yup, crash on closure is 100% reproducible. Let me file it as https://jira.zephyrproject.org/browse/ZEP-2423

pfalcon commented:
@andrewboie :

> Otherwise, if an issue is reproducible in the master branch, file a bug...

It's not reproducible in the master branch, that's why filing a Jira ticket doesn't feel right. I'm giving feedback to @mbolivar to the series of patches he's working on, of which this one is only a part, and actually beyond that, of what steps are made and issues are seen on bringing up BLE 6lowpan connectivity 96b_carbon. It started here on Github with patch series and feel like natural to discuss on github too. (I understand desire to keep epics/userstories and other PM activities in Jira; no, I don't see a big issue with having bug and PM trackers separate).

pfalcon commented:
https://github.com/zephyrproject-rtos/zephyr/pull/877 makes ipsp usable with `telnet fe80::c103:b6ff:fea2:bd03%bt0 4242` for me.

(I probably should add that some of the recent comments are more like a reply to https://github.com/mbolivar/zephyr/commit/78e980c9e586c16da50894d81980f3d461e3dbdf#commitcomment-23201065 , maybe I even should have posted them there, but then nobody else but us 2 would have a chance to read them, at least here, they can be found by github intra-project search).

Commit:
Bluetooth: Kconfig: delete stray double quote

Trivial fix.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
boards: 96b_carbon_nrf51: provide BT SPI slave configuration

Add board-specific CONFIG_BT_CONTROLLER_TO_HOST_SPI_* values.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
boards: 96b_carbon: provide BT SPI master configuration

On 96Boards Carbon, Bluetooth is provided by a secondary nRF51 chip
connected to SPI_1, so enable this peripheral and its driver when BT
is selected.

Similarly, provide BT_SPI_* configs to integrate with the BT HCI SPI
driver. The files these config values apply to only get built when
CONFIG_BT=y, but this configuration can't be handled in the "if BT"
section in 96b_carbon's Kconfig.defconfig.  This is because BT_SPI is
a choice value, and thus doesn't support a default setting.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>
Signed-off-by: Amit Kucheria <amit.kucheria@linaro.org>
Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>

Commit:
Bluetooth: ipsp: use sys_log

Various of the printk messages in this sample app don't have
newlines. Let's fix the output by moving to SYS_LOG_xxx() macros
instead, which don't need them anyway.  This also makes it easier to
tell errors apart from other messages, etc.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

mbolivar commented:
@pfalcon @dbkinder @jhedberg @cvinayak 

Now that https://github.com/zephyrproject-rtos/zephyr/pull/832 is merged, I've rebased and expanded this pull request. Main changes:

- Make Bluetooth HCI SPI changes needed to support controller based on new SPI API, for 96b_carbon and other STM32 targets
- Add missing configuration to set this up on 96b_carbon / 96b_carbon_nrf51
- Add a fix to the Bluetooth ipsp sample
- Add documentation for combining BT ipsp on 96b_carbon with hci_spi on 96b_carbon_nrf51

Could you please review?

mbolivar commented:
Updated to fix Sphinx target error; diff:

```
diff --git a/boards/arm/96b_carbon/doc/96b_carbon.rst b/boards/arm/96b_carbon/doc/96b_carbon.rst
index 4b5fde556..830d658d4 100644
--- a/boards/arm/96b_carbon/doc/96b_carbon.rst
+++ b/boards/arm/96b_carbon/doc/96b_carbon.rst
@@ -317,6 +317,8 @@ terminal:
 
    Hello World! arm
 
+.. _96b_carbon_verify_bluetooth:
+
 Verifying Bluetooth Functionality
 ---------------------------------
 
```

sjanc commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,295 @@
+			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+
+	/* Send a vendor event to announce that the slave is initialized */
+	buf = net_buf_alloc(&cmd_tx_pool, K_FOREVER);
```
this will always return valid net_buf

mbolivar replied:
Indeed.

sjanc commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,295 @@
+		case HCI_CMD:
+			memcpy(&cmd_hdr, &rxmsg[1], sizeof(cmd_hdr));
+
+			buf = net_buf_alloc(&cmd_tx_pool, K_NO_WAIT);
```
(maybe a stupid question) this is thread context so why not use K_FOREVER?

rsalveti replied:
Here it is just to avoid locking up the thread (similarly how it is done with the other hci_* samples), but I'm also not completely sure what would be the best way to handle such situation here, maybe @jhedberg can give us a hand here.

mbolivar commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,295 @@
+	do {
+		ret = spi_transceive(spi_hci_dev, header_slave, 5,
+				     header_master, 5);
+		if (ret < 0) {
```
Nice catch. Will fix this here, below, and recompile with -Wall -Wextra to make sure nothing else like this is around in this app.

mbolivar commented on samples/bluetooth/hci_spi/src/main.c:
```diff
@@ -0,0 +1,295 @@
+			SYS_LOG_ERR("SPI transceive error: %d", ret);
+		}
+
+		switch (rxmsg[PACKET_TYPE]) {
```
I'll add the missing continue in the error path.

pfalcon commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -10,6 +10,39 @@ Zephyr applications use the 96b_carbon configuration to run on the 96Boards
 
+.. note::
+
+   The nRF51832 chip on this board is supported by Zephyr using
```
If there's a suspicion of confusion (below), why not start with telling that the physical board has 2 chips (even mention something like "it's essentially two-core", people will love that)?

mbolivar replied:
That is already written above:

"It is based on the STMicroelectronics STM32F401RET Cortex-M4  CPU and also contains a nRF51832 chip connected over SPI for BLE connectivity."

mbolivar commented on samples/bluetooth/hci_spi/README.rst:
```diff
@@ -0,0 +1,46 @@
+Requirements
+************
+
+* A board with SPI slave, GPIO and BLE support
```
I'll add a period, but since this is not a full sentence, I think it's up to style whether it's "missing".

mbolivar replied:
... and while I'd love to add even more 96Boards branding into the mainline README for a generic application, I'm not sure that's appropriate for the high-level requirements, especially since the board-level documentation is going to feature prominent references to this app.

pfalcon replied:
> but since this is not a full sentence

And why it's not a full sentence? Text consists of paragraphs, paragraphs consist of sentences. ;-)

pfalcon replied:
> while I'd love to add even more 96Boards branding

I personally don't see it as "branding". This is a niche app, which won't work for 95% of boards. Someone may even wonder if there's a board for which it works, and can find it out. Yes, there're other ways to find it out, but the point of docs (README in our case) is to provide enough information, not make people "reverse engineer" it.

My overall concern is that this is not the first and not the last niche app, and not providing clear and useful info for each means that we target towards obscurity and esotericity (like someone with a board X may try 2, 3, 4 apps, none working for their board and silent about which boards it may work with).

YMMV

mbolivar commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -10,6 +10,39 @@ Zephyr applications use the 96b_carbon configuration to run on the 96Boards
+   different chip, than the 96b_carbon configuration described on this
+   page**.
+
+   This can be confusing. The important things to understand are:
```
The situation which is confusing is that two things which Zephyr calls a "BOARD" are present on the same physical printed circuit board. I think this confusion is due to the terminology adopted by Zephyr, which usually makes sense but breaks down in edge cases like this. If @dbkinder also objects I will remove this, but I prefer to let it stand, as I think the confusion is being explained by this documentation rather than caused by it.

mbolivar commented:
Rebased and updated to incorporate most review comments as well as adding more error handling; diff:

```
diff --git a/samples/bluetooth/hci_spi/README.rst b/samples/bluetooth/hci_spi/README.rst
index ea65a197c..478829e1a 100644
--- a/samples/bluetooth/hci_spi/README.rst
+++ b/samples/bluetooth/hci_spi/README.rst
@@ -10,7 +10,7 @@ the Zephyr SPI HCI transport protocol (similar to BlueNRG).
 Requirements
 ************
 
-* A board with SPI slave, GPIO and BLE support
+* A board with SPI slave, GPIO and BLE support.
 
 Building and Running
 ********************
diff --git a/samples/bluetooth/hci_spi/src/main.c b/samples/bluetooth/hci_spi/src/main.c
index 7ca37aa60..c71baabfc 100644
--- a/samples/bluetooth/hci_spi/src/main.c
+++ b/samples/bluetooth/hci_spi/src/main.c
@@ -95,7 +95,7 @@ static inline int spi_send(struct net_buf *buf)
 	u8_t header_master[5] = { 0 };
 	u8_t header_slave[5] = { READY_NOW, SANITY_CHECK,
 				 0x00, 0x00, 0x00 };
-	u8_t ret;
+	int ret;
 
 	SYS_LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf),
 		    buf->len);
@@ -153,7 +153,11 @@ static void bt_tx_thread(void *p1, void *p2, void *p3)
 	struct net_buf *buf = NULL;
 	struct bt_hci_cmd_hdr cmd_hdr;
 	struct bt_hci_acl_hdr acl_hdr;
-	u8_t ret;
+	int ret;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
 
 	memset(&txmsg, 0xFF, SPI_MAX_MSG_LEN);
 
@@ -179,6 +183,7 @@ static void bt_tx_thread(void *p1, void *p2, void *p3)
 				     &rxmsg, SPI_MAX_MSG_LEN);
 		if (ret < 0) {
 			SYS_LOG_ERR("SPI transceive error: %d", ret);
+			continue;
 		}
 
 		switch (rxmsg[PACKET_TYPE]) {
@@ -234,6 +239,8 @@ static int hci_spi_init(struct device *unused)
 		.config = SPI_WORD(8),
 	};
 
+	ARG_UNUSED(unused);
+
 	SYS_LOG_DBG("");
 
 	spi_hci_dev = device_get_binding(SPI_HCI_DEV_NAME);
@@ -263,26 +270,35 @@ void main(void)
 	static K_FIFO_DEFINE(rx_queue);
 	struct bt_hci_evt_hdr *evt_hdr;
 	struct net_buf *buf;
+	k_tid_t tx_id;
 	int err;
 
 	SYS_LOG_DBG("Start");
 
-	bt_enable_raw(&rx_queue);
+	err = bt_enable_raw(&rx_queue);
+	if (err) {
+		SYS_LOG_ERR("bt_enable_raw: %d; aborting", err);
+		return;
+	}
 
 	/* Spawn the TX thread, which feeds cmds and data to the controller */
-	k_thread_create(&bt_tx_thread_data, bt_tx_thread_stack,
-			K_THREAD_STACK_SIZEOF(bt_tx_thread_stack), bt_tx_thread,
-			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+	tx_id = k_thread_create(&bt_tx_thread_data, bt_tx_thread_stack,
+				K_THREAD_STACK_SIZEOF(bt_tx_thread_stack),
+				bt_tx_thread, NULL, NULL, NULL, K_PRIO_COOP(7),
+				0, K_NO_WAIT);
 
 	/* Send a vendor event to announce that the slave is initialized */
 	buf = net_buf_alloc(&cmd_tx_pool, K_FOREVER);
-	if (buf) {
-		bt_buf_set_type(buf, BT_BUF_EVT);
-		evt_hdr = net_buf_add(buf, sizeof(*evt_hdr));
-		evt_hdr->evt = BT_HCI_EVT_VENDOR;
-		evt_hdr->len = 2;
-		net_buf_add_le16(buf, EVT_BLUE_INITIALIZED);
-		spi_send(buf);
+	bt_buf_set_type(buf, BT_BUF_EVT);
+	evt_hdr = net_buf_add(buf, sizeof(*evt_hdr));
+	evt_hdr->evt = BT_HCI_EVT_VENDOR;
+	evt_hdr->len = 2;
+	net_buf_add_le16(buf, EVT_BLUE_INITIALIZED);
+	err = spi_send(buf);
+	if (err) {
+		SYS_LOG_ERR("can't send initialization event; aborting");
+		k_thread_abort(tx_id);
+		return;
 	}
 
 	while (1) {
```

Pending review comments:

- Documentation updates suggested (@pfalcon, waiting feedback from @dbkinder) 
- Switching to K_FOREVER in the TX thread (@sjanc, feedback from @jhedberg requested by @rsalveti)

FWIW, I've tried these patches with K_NO_WAIT (matching the other samples) and with K_FOREVER (suggested by @sjanc) and neither seems to have serious problems. So my preference would be to keep this as-is and follow up with a bulk change of all the HCI apps to use K_FOREVER if there is a good reason to do so.

mbolivar commented:
> And why it's not a full sentence? Text consists of paragraphs, paragraphs consist of sentences. ;-)

It lacks a verb, for one thing :).

> I personally don't see it as "branding". This is a niche app, which won't work for 95% of boards. Someone may even wonder if there's a board for which it works, and can find it out. Yes, there're other ways to find it out, but the point of docs (README in our case) is to provide enough information, not make people "reverse engineer" it.

This patch series already includes a reference to 96b_carbon_nrf51 from this very file (README.rst), just lower down in the page. I don't think it makes sense to put this information again in the list of requirements.

Here it it is, for reference:

```
Refer to :ref:`bluetooth_setup` for general Bluetooth information, and
to :ref:`96b_carbon_nrf51_bluetooth` for instructions specific to the
96Boards Carbon board.
```

dbkinder commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -281,6 +317,51 @@ terminal:
+functionality on the board. For help on Zephyr applications
+in general, see :ref:`build_an_application`.
+
+1. Make sure you have flashed the nRF51 with the hci_spi sample
```
I'd drop the "Make sure..." part and just say, "Flash the nRF51..."

mbolivar replied:
Done

dbkinder commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -281,6 +317,51 @@ terminal:
+1. Make sure you have flashed the nRF51 with the hci_spi sample
+   application as described in :ref:`96b_carbon_nrf51_bluetooth`.
+
+2. Make sure you have dfu-util installed as described earlier on this
```
Same here, "Install the dfu-util flashing app, as described above."

mbolivar replied:
Done

dbkinder commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -281,6 +317,51 @@ terminal:
+
+     $ make BOARD=96b_carbon flash
+
+Following that, refer to the instructions in
```
Since this is another step, use one more numbered list item:

     5. Refer to the instructions in ...

mbolivar replied:
Done

dbkinder commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -281,6 +317,51 @@ terminal:
+Debugging
+=========
+
+The 96b_carbon can be debugged by installing 100 mil (0.1 inch) header
```
installing a 100 mil...

mbolivar replied:
Done

dbkinder commented on samples/bluetooth/hci_spi/README.rst:
```diff
@@ -0,0 +1,46 @@
+Bluetooth: HCI SPI
```
add a label on line one:

     .. _bluetooth_hci_spi:

mbolivar replied:
Done

dbkinder commented on boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst:
```diff
@@ -117,6 +119,31 @@ output to a currently unused pin connected to the STM32F401RET. Users
+
+.. warning::
+
+   **Do not flash the hci_spi application on the STM32F401RET chip on
```
Instead of telling them what not to do, tell them what to do.  Replace the warning text with:

     Be sure to flash the hci_spi application to the nRF51832 chip and not to  
     the main STM32F401RET chip.  While both chips are supported by Zephyr,
     the hci_spi application providing Bluetooth support, will only run on
     the nRF51832 chip.

mbolivar replied:
Done

dbkinder reviewed (changes requested):
suggested edits

dbkinder commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -10,6 +10,39 @@ Zephyr applications use the 96b_carbon configuration to run on the 96Boards
 Carbon hardware. It is based on the STMicroelectronics STM32F401RET Cortex-M4
 CPU and also contains a nRF51832 chip connected over SPI for BLE connectivity.
 
+.. note::
```
I'm leaning more toward's @pfalcon's comment on this one.  I also don't think a NOTE should be this long either.  So here's a suggested rewrite of this section (all the way down to the figure):

     The 96b_carbon board is built with two chips: an STMicroelectronics
     STM32F401RET Cortex-M4 CPU and an nRF51832 chip connected to
     the Cortex-M4 CPU over SPI for Bluetooth LE connectivity.  Even though
     both chips exist on the same physical board, they must be programmed
     separately:

     - The ``96b_carbon`` configuration is used when developing programs for
       the main chip on the board, the STM32F401RET. Users will likely want to
       write applications targeting this chip, using the ``96b_carbon``
       configuration, since it is connected to all of the breakout
       I/O headers.

     - The ``96b_carbon_nrf51`` configuration should be used for programming
       the secondary nRF51832 chip. Most users will likely not develop
       applications for this chip, since Zephyr already provides a 
       sample application that can be flashed onto the nRF51832
       to provide Bluetooth functionality to applications on the main
       STM32F401RET chip.

     For instructions on how to set up the nRF51832 to develop Bluetooth
     applications, see :ref:`96b_carbon_nrf51_bluetooth`.
 
     After you have flashed your nRF51, you can perform basic validation
     of this Bluetooth setup using the instructions
     :ref:`below<96b_carbon_verify_bluetooth>`.

mbolivar replied:
I've got one nitpick about this:

> The 96b_carbon board is built with two chips

This is misleading, since "96b_carbon" is the Zephyr "BOARD" for the STM32, not the board itself. I think it'd be better to say "96Boards Carbon" to emphasize the physical object.

dbkinder replied:
I agree.  Good catch.  Make it so. :)

mbolivar replied:
Done, thanks :)

mbolivar commented:
Updated with documentation changes; diff:

```
diff --git a/boards/arm/96b_carbon/doc/96b_carbon.rst b/boards/arm/96b_carbon/doc/96b_carbon.rst
index 830d658d4..fc59164cc 100644
--- a/boards/arm/96b_carbon/doc/96b_carbon.rst
+++ b/boards/arm/96b_carbon/doc/96b_carbon.rst
@@ -10,38 +10,31 @@ Zephyr applications use the 96b_carbon configuration to run on the 96Boards
 Carbon hardware. It is based on the STMicroelectronics STM32F401RET Cortex-M4
 CPU and also contains a nRF51832 chip connected over SPI for BLE connectivity.
 
-.. note::
-
-   The nRF51832 chip on this board is supported by Zephyr using
-   another configuration, 96b_carbon_nrf51; see
-   :ref:`96b_carbon_nrf51_board` for more details on this
-   configuration.
-
-   **Although both chips exist on the same physical board,
-   96b_carbon_nrf51 is a distinct Zephyr configuration, targeting a
-   different chip, than the 96b_carbon configuration described on this
-   page**.
-
-   This can be confusing. The important things to understand are:
-
-   1. 96b_carbon should be used when developing programs for the main
-      chip on the board, the STM32F401RET. Users will likely want to
-      write applications targeting this chip, using the 96b_carbon
-      configuration, as it is what is connected to all of the breakout
-      I/O headers.
-
-   2. 96b_carbon_nrf51 should be used for the secondary chip, the
-      nRF51832.  Most users will likely not develop applications for
-      this chip, since Zephyr already provides a sample application
-      that can be flashed onto the nRF51832 to provide Bluetooth
-      functionality to the STM32F401RET.
-
-   For instructions on how to set up the nRF51832 to develop Bluetooth
-   applications using 96b_carbon, see :ref:`96b_carbon_nrf51_bluetooth`.
-
-   After you have flashed your nRF51, you can perform basic validation
-   of this Bluetooth setup using instructions :ref:`below
-   <96b_carbon_verify_bluetooth>`.
+The 96Boards Carbon board is built with two chips: an STMicroelectronics
+STM32F401RET Cortex-M4 CPU and an nRF51832 chip connected to
+the Cortex-M4 CPU over SPI for Bluetooth LE connectivity.  Even though
+both chips exist on the same physical board, they must be programmed
+separately:
+
+- The ``96b_carbon`` configuration is used when developing programs for
+  the main chip on the board, the STM32F401RET. Users will likely want to
+  write applications targeting this chip, using the ``96b_carbon``
+  configuration, since it is connected to all of the breakout
+  I/O headers.
+
+- The ``96b_carbon_nrf51`` configuration should be used for programming
+  the secondary nRF51832 chip. Most users will likely not develop
+  applications for this chip, since Zephyr already provides a
+  sample application that can be flashed onto the nRF51832
+  to provide Bluetooth functionality to applications on the main
+  STM32F401RET chip.
+
+For instructions on how to set up the nRF51832 to develop Bluetooth
+applications, see :ref:`96b_carbon_nrf51_bluetooth`.
+
+After you have flashed your nRF51, you can perform basic validation
+of this Bluetooth setup using the instructions
+:ref:`below <96b_carbon_verify_bluetooth>`.
 
 .. figure:: img/96b-carbon-front.png
      :width: 487px
@@ -326,11 +319,10 @@ This section contains instructions for verifying basic Bluetooth
 functionality on the board. For help on Zephyr applications
 in general, see :ref:`build_an_application`.
 
-1. Make sure you have flashed the nRF51 with the hci_spi sample
-   application as described in :ref:`96b_carbon_nrf51_bluetooth`.
+1. Flash the nRF51 with the hci_spi sample application as described in
+   :ref:`96b_carbon_nrf51_bluetooth`.
 
-2. Make sure you have dfu-util installed as described earlier on this
-   page.
+2. Install the dfu-util flashing app, as described above.
 
 3. Build the ``samples/bluetooth/ipsp`` application for 96b_carbon::
 
@@ -345,8 +337,8 @@ in general, see :ref:`build_an_application`.
 
      $ make BOARD=96b_carbon flash
 
-Following that, refer to the instructions in
-:ref:`bluetooth-ipsp-sample` for how to verify functionality.
+5. Refer to the instructions in :ref:`bluetooth-ipsp-sample` for how
+   to verify functionality.
 
 Congratulations! Your 96Boards Carbon now has Bluetooth
 connectivity. Refer to :ref:`bluetooth` for additional information on
@@ -355,7 +347,7 @@ further Bluetooth application development.
 Debugging
 =========
 
-The 96b_carbon can be debugged by installing 100 mil (0.1 inch) header
+The 96b_carbon can be debugged by installing a 100 mil (0.1 inch) header
 into the header at the bottom right hand side of the board, and
 attaching an SWD debugger to the 3V3 (3.3V), GND, CLK, DIO, and RST
 pins on that header. Then apply power to the 96Boards Carbon via one
diff --git a/boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst b/boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst
index 16baefe2f..a1c18ab34 100644
--- a/boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst
+++ b/boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst
@@ -136,13 +136,10 @@ Zephyr application, see :ref:`build_an_application`.)
 
 .. warning::
 
-   **Do not flash the hci_spi application on the STM32F401RET chip on
-   the 96Boards Carbon.**
-
-   Both chips are supported by Zephyr and exist on the same physical
-   board hardware, but the hci_spi application must run on the nRF51
-   chip supported by the 96b_carbon_nrf51 Zephyr configuration
-   described on this page.
+   Be sure to flash the hci_spi application to the nRF51832 chip and
+   not to the main STM32F401RET chip.  While both chips are supported
+   by Zephyr, the hci_spi application providing Bluetooth support will
+   only run on the nRF51832 chip.
 
 References
 **********
diff --git a/samples/bluetooth/hci_spi/README.rst b/samples/bluetooth/hci_spi/README.rst
index 478829e1a..32843031b 100644
--- a/samples/bluetooth/hci_spi/README.rst
+++ b/samples/bluetooth/hci_spi/README.rst
@@ -1,3 +1,5 @@
+.. _bluetooth_hci_spi:
+
 Bluetooth: HCI SPI
 ##################
```

dbkinder reviewed (dismissed):
thanks!

dbkinder commented:
... anticipating the tests all pass :)

mbolivar commented:
@galak @nashif is this a shippable false negative? I see 500 passed, 0 errored / failed / skipped.

@cvinayak have I addressed your concerns?

pfalcon commented:
@mbolivar :

> is this a shippable false negative? I see 500 passed, 0 errored / failed / skipped.

Tests are split in parallel batches. If you click on "3267.2" tab, you'll see "Tests 499  / 1  / 0  / 0".

(I already complained on IRC about recent change of UI design of Shippable, makes everything with Shippable confusing again, as if we just started with it).

mbolivar commented:
> Tests are split in parallel batches. If you click on "3267.2" tab, you'll see "Tests 499 / 1 / 0 / 0".

Many thanks. The failing test is:

```
 Sanitycheck / qemu_nios2:tests/kernel/workq/work_queue/test / qemu_nios2/tests/kernel/workq/work_queue/test
```

I think this is a false positive. I re-ran sanitycheck locally to confirm no failure, and this PR doesn't touch anything that I can imagine would affect this test's result.

pfalcon commented:
> and this PR doesn't touch anything that I can imagine would affect this test's result.

So, please rebase it to retrigger build and erase this random failure ;-).

P.S. I didn't yet have a chance to re-test this patch against master, hope to do this on weekend.

mbolivar commented:
> So, please rebase it to retrigger build and erase this random failure ;-).

Yes, I've seen people use that workaround, but that won't help anyone else running into the same issue, so I'm flagging it to see if the flaky test needs to be disabled.

@galak @nashif for what it's worth (since the "just rebase it" or "close it and reopen it" tricks are not 100% intuitive), some bots can monitor the comment stream for a magic string or otherwise take input that retriggers testing. I looked around in the shippable trigger docs (http://docs.shippable.com/validate/trigger-job/) and PR building docs (http://docs.shippable.com/ci/pull-request-builds/). Seems like it might be possible to rig something like that together with shippable, but I didn't see anything obvious.

pfalcon reviewed (dismissed):
Tested on Carbon.

cvinayak commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -10,6 +10,32 @@ Zephyr applications use the 96b_carbon configuration to run on the 96Boards
 CPU and also contains a nRF51832 chip connected over SPI for BLE connectivity.
 
+The 96Boards Carbon board is built with two chips: an STMicroelectronics
+STM32F401RET Cortex-M4 CPU and an nRF51832 chip connected to
```
nRF51822  not nRF51832. Please correct all instances.

mbolivar replied:
Thanks for the catch -- looks like that error is being propagated from the first introduction of the board. I'll fix all the instances, not just those introduced in this series.

cvinayak commented on samples/bluetooth/hci_spi/prj.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_BT_HCI_RAW=y
+CONFIG_BT_MAX_CONN=20
+CONFIG_BT_TINYCRYPT_ECC=n
+CONFIG_BT_RX_BUF_LEN=255
```
This shall be 258 for a payload of 251 bytes. Do see Data Lengths of 251 configured, why 255 bytes?

rsalveti replied:
BT_BUF_RX_SIZE is defined as CONFIG_BT_HCI_RESERVE + CONFIG_BT_RX_BUF_LEN, and HCI_RESERVE is 0 in this case. That value gets used by hci_raw when defining hci_rx_pool and hci_spi when defining cmd_tx_pool, so unclear how that that value can be larger than defined.

The size restriction we currently have is when calling spi_transceive, which is max 255, which is why this config was set to 255 as well.

carlescufi commented on boards/arm/96b_carbon/doc/96b_carbon.rst:
```diff
@@ -281,6 +310,50 @@ terminal:
+
+2. Install the dfu-util flashing app, as described above.
+
+3. Build the ``samples/bluetooth/ipsp`` application for 96b_carbon::
```
Wouldn't it be simpler to use the samples/bluetooth/beacon sample as a quick sanity test and then ask the user to simply scan the beacon with a phone or computer?

pfalcon replied:
I'm not sure if that would be simpler or not (I'm trying few random BT samples, and have to "reverse engineer" how they work), but IPSP definitely offers more functionality to the user (whole IP networking), so I'm +1 for keeping it.

carlescufi replied:
Right but we're talking sanity checking Bluetooth functionality here, and IPSP requires a counterpart to actually connect to the board and establish an IPSP connection, which essentially today means a Linux box and a few cryptic command-line commands. Users of the carbon board that are not interested in IP connectivity but want to use it for simple BLE functionality could do with an easier way to test it

pfalcon replied:
Right, then what should do users who just want a normal, conventional networking? Without guide thru those obscure commands, they will simply never get to it.

carlescufi replied:
I would put the list of the obscure commands in the IPSP sample README, and link from here to it in a separate section of the 96b_carbon.rst, not the one that describes how to quickly check if the Bluetooth controller is up and running

pfalcon replied:
Well, I agree that reusable doc parts should be factored out and linked to from the relevant places. Just as in general, docs need more work. But in the meantime, Networking/BLE support for 96b_carbon is in works for ~1year, and a lot of people wait for it exactly as it allows networking connectivity. So perhaps, docs refactoring can be left for another patch (this one is already big at 8 commits).

carlescufi replied:
Fair enough. Let's leave this for a follow-up patch and get this merged, agreed.

carlescufi commented:
@mbolivar I just posted a PR to rename BT_CONTROLLER to BT_CTLR. Sorry for the inconvenience if mine gets merged before yours: https://github.com/zephyrproject-rtos/zephyr/pull/1106

mbolivar commented:
> @mbolivar I just posted a PR to rename BT_CONTROLLER to BT_CTLR. Sorry for the inconvenience if mine gets merged before yours

Thanks for the heads up!

nashif commented:
> @galak @nashif for what it's worth (since the "just rebase it" or "close it and reopen it" tricks are not 100% intuitive), some bots can monitor the comment stream for a magic string or otherwise take input that retriggers testing. I looked around in the shippable trigger docs (http://docs.shippable.com/validate/trigger-job/) and PR building docs (http://docs.shippable.com/ci/pull-request-builds/). Seems like it might be possible to rig something like that together with shippable, but I didn't see anything obvious.

we have an issue open about this: #779

nashif commented:
this needs rebasing now anyways, so please rebase and we might get rid off the false positive

mbolivar commented:
Rebased and updated. Diff: https://gist.github.com/mbolivar/729dc895e1d36d272491d2efa9ad53e3

@cvinayak, I've left CONFIG_BT_RX_BUF_LEN=255. OK with you? I think that's the last potential issue.

mbolivar commented:
> we have an issue open about this: #779

Thanks for the hint. I'll watch this one since I've seen GH issues are getting shut down: https://jira.zephyrproject.org/browse/INF-134

cvinayak commented:
@mbolivar CONFIG_BT_RX_BUF_LEN=255 is ok, with the fact that its the restrictions over SPI.

cvinayak reviewed (approved):
Looks good to me
